### DISCUSSION TITLE
How to convert numpy arrays into Const() lookup tables

### DISCUSSION THREAD

**Discussion by jrmoserbaltimore** [amaranth-lang/amaranth#1053] [Q&A]
What's the best way to convert a Numpy array into Const() lookup table?

Let's say you want to generate 8 mipmapped 24-bit integer band-limited sawtooth wave tables on power of 2 base frequencies from 16Hz to 2048Hz.  Because these only contain odd-order harmonics, you only need to generate half the waveform (you need to roll around and invert direction to use it, i.e. 0, 1, 2, 3, 0, -1, -2, -3... and multiply the values at negative indexes by -1).

Doing this with Numpy using an 8x1024 array and converting the `float64` to `uint32` (each half is entirely positive or negative) containing values scaled to 24-bit integer representation is doable pretty simply:

```python
from amaranth import *
import numpy as np
from numpy import sin, uint32, float64, power
tau = 2*np.pi

temp_ramp_tables = np.empty((8,1024), dtype=float64)

for i in range(0,8):
    base_f0 = power(2, i+4)
    n_harmonics = np.floor(18000/base_f0)
    # Generate 1024 samples normalized to 1/2 waveform (hence /2*tau))
    # and including all harmonics up to 18kHz
    h = np.arange(1, n_harmonics+1)
    temp_ramp_tables[i] = np.sum(sin(h * base_f0 / (2*tau)) / h)

# compress peak-to-peak relative to the largest amplitude
temp_ramp_tables /= np.max(temp_ramp_tables)
# convert to positive half of unsigned 24-bit range
ramp_tables = temp_ramp_tables*(power(2,23) - 1).astype(uint32)
del temp_ramp_tables
```

The question is how to convert to a table for Amaranth?  The best I can come up with is the below change:

```python
# convert to positive half of unsigned 24-bit range
ramp_tables = [
               [Const(x, unsigned(23)) for x in temp_ramp_tables[y]*(power(2,23) - 1).astype(uint32)]
               for y in range(0,8)
              ]
del temp_ramp_tables
```

Is this the best way?  Is there a better way?

**Comment by whitequark** [comment]
> ```python
> [Const(x, unsigned(23)) for x in temp_ramp_tables[y]*(power(2,23) - 1).astype(uint32)]
> ```

I would say this is the best way, yeah
